include/deal.II/dofs/dof_renumbering.h

```diff
@@ -1264,6 +1273,62 @@ namespace DoFRenumbering
    * @}
    */
 
+  /**
+   * @name Numberings for better performance with the MatrixFree infrastructure
+   * @{
+   */
+
+  /**
+   * Sort DoFs by their appearance in matrix-free loops in order to increase
+   * data locality. More specifically, this renumbering strategy will group
+   * DoFs touched only on a single group of cells (as traversed by a
+   * MatrixFree::cell_loop()) nearby, whereas DoFs touched far apart through
+   * the loop over cells are grouped separately. Obviously, also unknowns that
+   * are subject to ghost exchange will be treated as those with far reach,
+   * and will be grouped accordingly. Currently, this function only works for
+   * finite elements of type FE_Q as well as systems of a single FE_Q element.
+   *
+   * This function needs to be provided with an appropriate
+   * MatrixFree::AdditionalData structure that indicates whether level DoFs or
+   * active DoFs are to be renumbered through the variable
+   * MatrixFree::AdditionalData::mg_level, and with the options that determine
+   * the order cells are traversed in a matrix-free loop.
+   *
+   * Note that the argument `matrix_free` does not need to be initialized when
+   * calling this function. In case it is already set up, the unknowns stored
+   * in that data will be used. If it is not set up, the
+   * MatrixFree::AdditionalData is instead used to construct an object with
+   * the appropriate indices, which are then processed. In both cases, the
+   * content of MatrixFree is left untouched. However, both MatrixFree and the
+   * associated AffineConstraints object need to be computed again after
+   * changing the numbers of the degrees of freedom.
+   */
+  template <int dim, int spacedim, typename Number, typename MatrixFreeType>
```

Comment:
In the new variant, I can spell out the exact `MatrixFree` class in that variant. However, for `AdditionalData` I opted to go with an `AdditionalDataType` because otherwise the compiler can't deduce the template types, so the user has to manually write them in the code (which I find very annoying).